[MRG+1] Add Huber Estimator to sklearn linear models

MechCoder opened the pull request:
Add robust regression model that filters outliers based on http://statweb.stanford.edu/~owen/reports/hhu.pdf
- [x] Add fix for random OverflowErrors.
- [x] Add documentation to the helper function
- [x] Add extensive testing
- [x] Add narrative docs
- [x] Add example
- [x] Support for sparse data
- [x] Support sample_weights

MechCoder committed:
Huber regressor

Add gradient calculation in _huber_loss_and_gradient

Add tests to check the correctness of the loss and gradient

Fix for old scipy

Add parameter sigma for robust linear regression

Add gradient formula to robust _huber_loss_and_gradient

Add fit_intercept option and fix tests

Add docs to HuberRegressor and the helper functions

Add example demonstrating ridge_regression vs huber_regression

Add sample_weight implementation

Add scaling invariant huber test

Remove exp and add bounds to fmin_l_bfgs_b

Add sparse data support

Add more tests and refactoring of code

Add narrative docs

review huber regressor

Minor additions to docs and tests

Minor fixes that deals with dealing with NaN values in targets
and old verions of SciPy and NumPy

Add HuberRegressor to robust estimator

Refactored computation of gradient and make docs render properly

Temp

Remove float64 dtype conversion

trivial optimizations and add a note about R

Remove sample_weights special_casing

address @amueller comments

jmschrei commented:
HuberLoss is already implemented in ensemble/GradientBoosting.py. I feel like we need to have a more centralized loss module, rather than reimplementing them when needed. What are your thoughts?

MechCoder commented:
I agree with you. Has there been an ongoing discussion already that I've missed?

jmschrei commented:
#5044 has some comments about it, but no consensus has been reached yet.

MechCoder commented:
I just looked through the code in `sklearn.ensemble.gradient_boosting` . I do agree that the code refactoring has to be done but I'm not sure if that should block this PR.

1. The current code in `HuberLoss` does not take into account regularization it looks like the alpha is not the regularization alpha, but a way to calculate the epsilon value)
2. I can also wrap my methods with calls to HuberLoss(n_c)(y, pred) and negative_gradient(). But there is some code duplication which might be expensive while calculating the loss and the gradient.

What do you think?

MechCoder commented:
Btw, this fixes https://github.com/scikit-learn/scikit-learn/issues/4990

jmschrei commented:
If there is a way of doing it currently, it may be worth benchmarking it to see how expensive it is. If there is a significant slowdown, you should go ahead if it solves a pressing issue.

mblondel commented:
> I feel like we need to have a more centralized loss module, rather than reimplementing them when needed. What are your thoughts?

I don't think this would be so useful. This PR uses the gradient w.r.t. linear model parameters (size is n_features). GradientBoosting uses the gradient with respect to the predictions (size is n_samples). So the code is different. The one part that could be possibly shared is computing the objective value but the regularizer is different.

jmschrei commented:
Okay, then!

dbtsai commented:
I am working on robust regression for Spark's MLlib project based on Prof. Art Owen's paper, [A robust hybrid of lasso and ridge regression](http://statweb.stanford.edu/~owen/reports/hhu.pdf ). In MLlib/Breeze, since we don't support L-BFGS-B while the scaling factor in Eq.(6) \sigma has to be >= 0, we're going to replace it by \exp(\sigma). However, the second derivative of Huber loss is not continuous, this will cause some stability issue since L-BFGS requires it for guaranteed convergence. The workaround I'm going to implement will be [Pseudo-Huber loss function](https://en.wikipedia.org/wiki/Huber_loss#Pseudo-Huber_loss_function) which can be used as a smooth approximation of the Huber loss function, and ensures that derivatives are continuous for all degrees.

BTW, in robust regression, the scaling factor \sigma has to be estimated as well, and this is \epsilon in your case. This value can not be a constant. Imagine that, when the optimization is just started with some initial condition, if the initial guess is not good, then most of the training instances will be treated as outliners. As a result, \epsilon will be larger, but will be one of the parameters that will be estimated. See the details in Prof. Art Owen's paper in section 4. Thanks.

MechCoder commented:
Thanks for the comment and the link to the paper. (And it comes at a time when my benchmarks weren't looking too great)

Previously I was using grid search to find out the optimal value of epsilon, but it always corresponded to the lowest value of epsilon. (i,e assuming both X and y are centered and scaled)

To describe the paper in a short manner

1. It seems that the epsilon over here corresponds more to the the parameter M in the paper which is said to be fixed at 1.35.
2. In addition to that the gradient term `y - X*w -c` is scaled down by a factor sigma which makes the algorithm scaling independent.
3. And since the new function as described in 8 is jointly convex, we could optimize sigma together with the coefficients right?

This apart from the fact that the Hessian is not continuous hence we might need to use a pseudo=convex function to ensure convergene right?

dbtsai commented:
You are right. But you may want to replace \sigma to \exp(\alpha) so you don't need to have the condition that \sigma > 0. In theory, the hssian is not continuous, so LBFGS may not work well. But I don't know the exact impact on this. We may need to do some benchmark on this.

dbtsai commented:
Also, for Pseudo-Huber loss, there is no proof that it will be jointly convex with \sigma. Although I guess it will if we go through the proof.

MechCoder commented:
Great I'll try two things.

Change the present loss function to accomodate minimizing sigma. (and after that)
And after that I can try the pseudo Huber loss to check if there is any noticeable change in convergence (etc).

dbtsai commented:
Sounds great. Let me know the result, so I can learn from you when I implement this in Spark. Thanks.

MechCoder commented:
@dbtsai I've made changes to the loss function, but I'm not getting good results. Could you please verify if the loss function is right?

MechCoder commented:
Good results meaning that this is the plot that I generated from the coefficients :/

![figure_1](https://cloud.githubusercontent.com/assets/1867024/10174455/cc929e34-66b7-11e5-97af-23725e93fea4.png)

The red line is the one from the HuberRegressor and the green is the RidgeRegression. As you can clearly see that it is not what it is supposed to look like.

dbtsai commented:
@MechCoder I will compare the note I have in my home tonight. What do you mean you don't get good result? How do you test it? Also, when `\epsilon` is large, does it converge to normal LiR? Thanks.

dbtsai commented:
Can you try to make `\epsilon` very large and see if you can reproduce the RidgeRegression?

MechCoder commented:
I tried that as well, but it seems that epsilon has almost no effect (for both very high and very low values ) since

since |(y - X'w) / exp(sigma)| < M 

|y - X'w| < M*exp(sigma) . So the limit will change for every iteration the loss function is called no?

MechCoder commented:
Just in case you are interested with the plot generation

(https://gist.github.com/MechCoder/8205f0fce4395a9ab907)

MechCoder commented:
oops seems like I made a mistake. Just a second.

dbtsai commented:
Here is the note I compute dL/d\sigma Let's compare if we get the same formula. 

![img_0271](https://cloud.githubusercontent.com/assets/1134574/10177172/8fe455c2-66ad-11e5-8def-e5d50615f813.JPG)

MechCoder commented:
@dbtsai I modified the loss function just before your comment :P . I have commented out the gradient out for now and I set `approx_grad=True` in `fmin_l_bfgs_b`.

I just wanted to have an approximate idea if the loss function is correct. After making the changes to the loss function (it should be clearer now), I am able to replicate the behavior of ridge for high values of epsilon. (note that the lines coincide)

![figure_1](https://cloud.githubusercontent.com/assets/1867024/10177266/45360fa6-66c7-11e5-8d9c-9d7e0bf74bd3.png)

MechCoder commented:
I will readd the gradient and verify it from your note in a bit.

dbtsai commented:
Cool. It looks nice! How about small `\epsilon`? Does it help to filter out the outlier?

dbtsai commented:
Also, can you try smooth huber loss as well? Thanks.

MechCoder commented:
I just derived the gradient function and added it back. I verified from the formula in your note that it is correct. Anyhow, could you also check from the code if the gradient is right?

I can play with the code tomorrow with different values of epsilon and different data and check the performance.

After that we can check about the smooth huber loss. 

Thanks.

dbtsai commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,93 @@
+    grad[-1] -= n_outliers * epsilon**2 * exp(sigma)
+    grad[-1] -= squared_loss
+
+    return X.shape[0] * exp(sigma) + squared_loss + outlier_loss + alpha * np.dot(w, w), grad
```
In fact, I don't know how to handle regularization easily since that term should be \alpha / exp(sigma) in order to get the same result compared with the one without robust. As a result, that term will contribute to dL/da

dbtsai commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,93 @@
+    outliers_true_pos = np.logical_and(linear_loss >= 0, outliers_true)
+    outliers_true_neg = np.logical_and(linear_loss < 0, outliers_true)
+    grad[:n_features] -= epsilon * X[outliers_true_pos, :].sum(axis=0)
+    grad[:n_features] += epsilon * X[outliers_true_neg, :].sum(axis=0)
```
I thought this is 2 * epsilon?

MechCoder replied:
Right.

dbtsai commented:
Oh, it seems that you are mixing two notations. In general, I prefer to have `/ 2` in the square loss, but it seems that you are mixing both if I read it correctly. 

1) Since \epsilon has to be `>= 1`, unless we handle it based on eq. (14). But this is rare case, so let's make it working for `M >= 1`. 
2) Regularization may not work like this as the comment. Let's work on this later.
3) It's easier to check in python. Can you add couple strong outliers in your synthetic data, and tune M from 1.0 to infinity to see how it perform? When M is very large, we should get the normal ridge regression. 
4) Although it's a scaling constant, I wonder how the robust regression will perform with/without dividing by two in the square loss. 

Thanks.

MechCoder commented:
I don't know what exactly you mean by "mixing" but if you are referring to the 2 factor in the linear loss, I have added it .I prefer to have 0.5 in the squared loss as well, but I kept it as the actual squared loss since it is mentioned like that in equation 5]

1. Right.

2. I'm unable to understand completely why or why not this may be true. Are you suggesting that w is somehow scaled down by `exp(sigma)` and then we need to scale `w` by `exp(sigma)` to penalize it. However also note that the improved penalty term in equation 7 does not contain any sigma, it only contains `tau`, which might mean the sigma should not influence the penalty term.

3. Yes about to do that.

We should also replace `a + b*exp(sigma)` to `a*exp(-sigma) + b` wherever possible, since it will overflow for large values of sigma.

MechCoder commented:
Anyway for now, I shall plot with lambda = 0, and we can investigate further.

MechCoder commented:
Finally some results after ten days of frustration.
 
These are the plots that I get with some noisy data (data from astroML.datasets).

![figure_1](https://cloud.githubusercontent.com/assets/1867024/10204474/ee0e1a5c-6789-11e5-8b5d-e5bc688b130c.png)

dbtsai commented:
Yes, the mixing is the 0.5 in the squared loss. But since in Eq8, there is no 0.5 in the square error, let's put 0.5 in the first term of `n * \sigma / 2` to make it exactly the same as paper for now. 

2) I am not sure about this yet. How about we make it working without regularization first, and we think about how to do it properly.

I don't think that replacing  `a + b*exp(sigma)` to `a*exp(-sigma) + b` will minimize the same loss.

dbtsai commented:
Great. The plot looks awesome. But I only see three lines in the plot instead of five.

MechCoder commented:
> But since in Eq8, there is no 0.5 in the square error ..

I'm not sure what you mean. I've removed the 0.5 term from the squared error to match the equation. https://github.com/scikit-learn/scikit-learn/pull/5291/files#diff-7416ccedd45a5840c67ff7877d24e1ceR36 and the grad has a `2` term because of this (https://github.com/scikit-learn/scikit-learn/pull/5291/files#diff-7416ccedd45a5840c67ff7877d24e1ceR43) . Now it's just `n * exp(sigma) + the huber loss term` which matches the equation. Are you referring to this or something else?

> I don't think that replacing a + b*exp(sigma) to a*exp(-sigma) + b

Ah yes, I keep forgetting that sigma is also a param :/

dbtsai commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,93 @@
+    # This is equal to (2 * M * |y - X'w / exp(sigma)| - M**2)*exp(sigma)
+    n_outliers = np.count_nonzero(outliers_true)
+    outliers = abs_linear_loss[outliers_true]
+    outlier_loss = 2 * epsilon * np.sum(outliers) - exp(sigma) * n_outliers * epsilon**2
```
I thought we're going to put 0.5 here.

dbtsai commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,93 @@
+    # Calculate the quadratic loss due to the non-outliers.-
+    # This is equal to |(y - X'w)**2 / exp(2*sigma)|*exp(sigma)
+    non_outliers = linear_loss[~outliers_true]
+    squared_loss = exp(-sigma) * np.dot(non_outliers, non_outliers)
```
ditto.

dbtsai commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,93 @@
+    grad = np.zeros(n_features + 1)
+
+    # Gradient due to the squared loss.
+    grad[:n_features] = 2 * exp(-sigma) * np.dot(non_outliers, -X[~outliers_true, :])
```
Then this 2 can be removed.

dbtsai commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,93 @@
+    # Gradient due to the linear loss.
+    outliers_true_pos = np.logical_and(linear_loss >= 0, outliers_true)
+    outliers_true_neg = np.logical_and(linear_loss < 0, outliers_true)
+    grad[:n_features] -= 2 * epsilon * X[outliers_true_pos, :].sum(axis=0)
```
Remove the 2 here

dbtsai commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,93 @@
+    grad[:n_features] += alpha * 2 * w
+
+    # Gradient due to sigma.
+    grad[-1] = n_samples * exp(sigma)
```
Since we add 0.5 in the square loss, we also have to add 0.5 here to make the loss effectively the same as paper.

```python
grad[-1] = 0.5 * n_samples * exp(sigma)
grad[-1] -= 0.5 * n_outliers * epsilon**2 * exp(sigma)
grad[-1] -= squared_loss

return 0.5 * X.shape[0] * exp(sigma) + squared_loss + outlier_loss + alpha * np.dot(w, w), grad
```

MechCoder replied:
True, but that only holds if I added 0.5 in the squared loss in the first place :P

dbtsai commented:
Oh, just for notation, I prefer to optimize

```
loss = \sigma / 2 + 0.5 ((y - wx)/\sigma)^2 \sigma  when y-wx <= \sigma M
        = \sigma / 2 + 0.5 (2M | y - wx| - \sigma M^2)   when y-wx > \sigma M
```
Thanks.

MechCoder commented:
OK. But scaling up or down the loss by a constant factor should not matter right?

I have done exactly as you have done multiplied by a factor of 2, i.e

```
loss = \sigma + ((y - wx)/\sigma)^2 \sigma  when y-wx <= \sigma M
        = \sigma + 0.5 (2M | y - wx| - \sigma M^2)   when y-wx > \sigma M
```

which explains the 2 factor in the gradient. I'd prefer it to keep it this way unless you feel it is absolutely necessary to change it.

MechCoder commented:
New plots, I narrowed the range to 3.5 since values around that were similar to ridge regression. You can see that the line approaches the ridge regression line.

![figure_1](https://cloud.githubusercontent.com/assets/1867024/10205591/12604fb4-6790-11e5-9df4-952f5200e60b.png)

dbtsai commented:
Oh, in Spark version, I will probably have 0.5 in the square loss to keep it consistence with other code; however, it's not necessary. 

I am wondering

 ```
loss = \sigma / 2 + 0.5 ((y - wx)/\sigma)^2 \sigma  when y-wx <= \sigma M
        = \sigma / 2 + 0.5 (2M | y - wx| - \sigma M^2)   when y-wx > \sigma M
```
and
```
loss = \sigma  + 0.5 ((y - wx)/\sigma)^2 \sigma  when y-wx <= \sigma M
        = \sigma  + 0.5 (2M | y - wx| - \sigma M^2)   when y-wx > \sigma M
```
will give very different result. My gut feelings is that we will get similar result. But let's follow the paper.

BTW, the result looks awesome. Do you know how 95% translates to M=1.35? Wondering if setting percentage for users will be easier.

dbtsai commented:
Also, can you compare the smooth version of huber loss as well? You don't have to implement the gradient to compare. Thanks.

MechCoder commented:
> Do you know how 95% translates to M=1.35?

I thought they assumed the noise to be normally distributed, hence assumed that 95% of the noise lies within 1.35 units from the mean. But from the graph of the normal distribution it seems that this value should be greater than 1.5 and less than 2, so my thoughts are wrong.

> (If you can verify this, it will be great since it's slower to iterate in Spark

I verified by multiplying a factor of 0.5 to the loss and keeping the `sigma` term as it is.
```
loss = \sigma  + 0.5 ((y - wx)/\sigma)^2 \sigma  when y-wx <= \sigma M
        = \sigma  + 0.5 (2M | y - wx| - \sigma M^2)   when y-wx > \sigma M
```

I get this graph.

![figure_1](https://cloud.githubusercontent.com/assets/1867024/10229092/57caeae6-6844-11e5-8a34-a7b42a759cf8.png) which shows variation from the original graph. btw, why does replacing `\sigma/2` with `\sigma` make it easier in Spark?

I'll post results with the pesudo huber loss in a while.

dbtsai commented:
I kind of know how to compute 1.35 for 95% robustness. I will post the equation this weekend. In Spark, I will let users configure the percentage of robustness instead of the magic number like M. This 1.35 is computed from Huber loss function, so if other robust loss is used, it will be different.

Just wondering. I will put 0.5 in square loss, so I have to put it in \sigma term. I will implement the one in the paper in this case. 

Thanks. Looking forward to the result to smooth huber loss.

MechCoder commented:
I get these values for the pseudo-huber function.

    def _pseudo_huber_loss_and_gradient(w, X, y, epsilon, alpha):
        sigma = w[-1]
        w = w[:-1]

        residual = (y - np.dot(X, w))**2
        residual /= epsilon**2
        residual *= exp(-2*sigma)
        residual += 1
        residual = np.sqrt(residual, residual)
        residual -= 1
        residual *= epsilon**2 * exp(sigma)
        return exp(sigma) * X.shape[0] + np.sum(residual) 
![pseudo_huber](https://cloud.githubusercontent.com/assets/1867024/10231871/38f63a9e-6853-11e5-9340-fe4ae29c42f6.png)

MechCoder commented:
The `d['warnflag]` from optimize.lbfgs says that the loss from both the functions converge.

Also I think we are free to use whatever loss function we want, to optimize sigma as long as it is convex. (Theorem 1).

What does `retaining 95% statistical efficiency` exactly mean?

MechCoder commented:
From the pseudo huber function it takes a much higher value of epsilon (for this toy data) to resemble ridge regression.

![pseudo_huber](https://cloud.githubusercontent.com/assets/1867024/10232050/4b2378fc-6854-11e5-8364-e8bd7c91a60c.png)

dbtsai commented:
The potential issue is not the Hessian of huber is not continuous, so it's not guaranteed that lbfgs will be converged. But lots of people still use lbfgs for this situation, and I don't think the issue will be very serious since it only happens when one instance is exactly on M * \sigma, and most/all of the instances will be not exactly on M * \sigma, so this will smooth out the discontinuity I guess. 

I will stick with Huber loss. For `retaining 95% statistical efficiency`, I will post another article explaining what is it, and how to compute it.

MechCoder commented:
Umm.. all right, I shall writing docs and extensive tests, so that this PR becomes merge worthy :)

MechCoder commented:
Oh by "we are free to use whatever loss function we want, to optimize sigma as long as it is convex."

I meant to address your concern about the joint convexity of sigma and w when using the pseudo-huber loss. (Not the concern about using lbfgs where the derivatives are non-continuous).

dbtsai commented:
Get you.

dbtsai commented:
BTW, which one will you use in the end? Huber or approximated Huber?

MechCoder commented:
For now, I think we can use just the huber function, since it gives fair results. However we can keep an issue open, if in the future and if proved that we have significant advantages of having the pseudo-huber, we can then use that.

dbtsai commented:
I figured today that subtracting the means and dividing by the variances may not be idea. Since they will be biased by the outliers, forcing the hyperplane on the center of features and response maybe cause issue. I think we should fit the intercept explicitly here, and/or estimate the mean and variance using Robust statistics.

MechCoder commented:
Yes, that is true. But I'm not sure we need to scale down explicitly by the std for preprocessing since that is taken care of by the scale factor.

dbtsai commented:
As you can see now, in your experiment, all the lines are crossing the same point which is the mean of x and y. Will be nice to do an experiment to learn the intercept with/without scaling the features. I know the scaling is kind of addressed by the scale factor, but it doesn't solve the issue that some of the features may have higher std.

MechCoder commented:
Now the loss is minimized for the intercept within the loss function. It would be great if you could verify that the loss and the gradient function is correct.

Before this commit:
![huber_intercept_old](https://cloud.githubusercontent.com/assets/1867024/10290348/9359e1cc-6b71-11e5-950e-7b2b5dd160bf.png)

You can see that all points pass through X_mean, y_mean since the intercept is set to (y_mean - X_mean * w)

After this commit:

![huber_intercept_new](https://cloud.githubusercontent.com/assets/1867024/10290415/e2565e4a-6b71-11e5-8e87-fe7560089e7d.png)

MechCoder commented:
@dbtsai You are completely right that the sigma parameter makes the loss function invariant to the scaling of y, but it might not have an effect if the different features in X are scaled to different factors.

Meanwhile I've added a TODO list on the top of the list, feel free to add if you think that if anything else is also important.

dbtsai commented:
Cool. The latest result makes more sense. I think by just scaling each variables with ANY scalers in features will not change the result if there is no regularization. So we can still divide by the variances but fit the intercept.

MechCoder commented:
Interesting. I'll try it out tomorrow.

MechCoder commented:
@dbtsai I added an example here (https://github.com/MechCoder/scikit-learn/commit/581df896917c4c08355fd5273d301443ec60a460) . Do you have any other suggestions for examples that would be useful for the end user?

dbtsai commented:
The examples look great for me. Do you try the scaling? At one point, I proofed that that should be invariance. Thanks.

MechCoder commented:
Great , I'll so that tomorrow. Lot of homework due :P

MechCoder commented:
@dbtsai I added a scaling invariant test which fails unfortunately :( Can you look at the test in the last commit to see if its alright?

Meanwhile, I think we should use something like `sigma(**2)`  instead of exp(sigma) is our goal is to just limit the scale to positive values, since exp(sigma) overflows very quickly, and I've got a number of `OverflowError`s while testing already.

MechCoder commented:
And by the way, I assume you have proved somewhere that if H is convex, then `H( eta / exp(sigma)) * exp(sigma)` is also jointly convex with eta and sigma? (The first derivative of H cancels out in the derivation). Because I just read the proof and it does not seem to be straight forward, I will have to work through while replacing sigma with other functions of sigma.

MechCoder commented:
Actually http://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.fmin_l_bfgs_b.html#scipy.optimize.fmin_l_bfgs_b has a bounds parameter and we can just pass the bounds as @amueller pointed out lol.

dbtsai commented:
I don't start to look through your code as I am very busy on work recently. With work on it and spark version durning this weekend. BTW, if you are interested in, maybe spark version can be our join work. 

Do you try bounds and compare with the exp trick? Thanks.

amueller commented:
can you check the reference? The current one seems to have only 60 citations.

dbtsai commented:
@amueller are you talking about the paper from Art Owen? He is a well reputed professor in this field at Stanford, and I think the citations are not high due to it's more statistical research not machine learning problem.

amueller commented:
I'm not refuting his credibility, I just think there might be another, better-cited paper by him. This is a well-known method, right?

dbtsai commented:
Replacing \sigma by \exp(\alpha) for making sure that \sigma is positive was my idea, and other than that, this PR pretty much implements his paper. He mentioned in the same paper that Peter J. Huber also came out with the same formula, I'll double check with the book, http://www.wiley.com/WileyCDA/WileyTitle/productCd-0470129905.html

MechCoder commented:
@dbtsai Just passing in the bounds to `fmin_lbfgs_b` seems to be faster and I get the same results.
(and also there is scale invariance now, irrespective of the scale applied to X and y. I've added tests for that)

Also, I am not sure one can replace F(w/sigma) with F(w/f(sigma)) (unless you have proved that the joint convexity holds if sigma is replaced with exp(sigma) as well, which involves a fair amount of symbol crunching).

amueller commented:
I'm even more confused by the reference now. "A robust hybrid of lasso and ridge regression" introduces a **Reverse Huber**. The original huber book is the right reference, it looks to me.

amueller commented:
see http://sanghv.com/download/Soft/Machine%20Learning,%20Artificial%20Intelligence,%20Mathematics%20eBooks/math/statistics/robust%20statistics%20%282nd,%202009%29.pdf for a totally not pirated version

amueller commented:
The interesting part is Section 7.7, in particular Theorem 7.9 on page 182, or pdf page 194. I didn't exactly follow the update he uses in 7.9 because there is quite a bit of notation to go through. It looks like he has some specific update for the scale factor, though, which might be faster than using l-bfgs on everything.

amueller commented:
Having a better parametrization (and a better name) then epsilon would be nice. Also: is alpha as in ridge or as in lasso? I guess as in lasso makes more sense? Have you tried it?

amueller commented:
can you test against an (R?) reference implementation?

dbtsai commented:
@MechCoder  What do you mean that `fmin_lbfgs_b` seems to be faster? Less iteration? Also, what do you do to make it scale invariance? Does `lbfgs` with exp trick and `lbfgs_b` give you the same result? Since in Spark, we don't have `lbfgs_b`, we need to use exp trick, and we have to prove that it's convex as well.

@amueller I think `Reverse Huber` was introduced by Art Owen for new regularization method, and this is not required in general in my understanding. It seems that after 7.8.1, he is proving a faster algorithm in convergency instead of putting everything in optimization, but I don't get the full picture yet. However, eq 7.118 intrigues me. In Art's paper, `a = 1` in 7.113, but in Huber's book, `a = (n - p) / n E(\chi)`. Do you know the idea here?

Thanks.

MechCoder commented:
Firstly, with the `exp` trick I am unable to prove scale invariance (i.e different samples get filtered as outliers). The test that I added failed.

In I just set bounds on sigma so that it does not become negative and follow the same approach as mentioned in the paper, i.e whatever is previously exp(sigma) is now just sigma.

dbtsai commented:
Interesting. I will do more investigation on this. BTW, breeze just implemented LBFGS-B not long ago. https://github.com/scalanlp/breeze/blame/master/math/src/main/scala/breeze/optimize/LBFGSB.scala

MechCoder commented:
I have updated the status to [MRG]

Would highly appreciate it if first passes could be given and if this could be covered during the sprint.

cc @amueller @agramfort

amueller commented:
@MechCoder sorry I'm swamped. Maybe during the sprint.

kastnerkyle commented on doc/modules/linear_model.rst:
```diff
@@ -1019,6 +1019,62 @@ considering only a random subset of all possible combinations.
+
+where
+
+.. math:: H(z)
```
H_m(z) to stay consistent with paper notation (eq. 5 subsection 2 has M, the eq above which only has H(z) has no M term). Also H_M above and here would be more consistent IMO

kastnerkyle commented on examples/linear_model/plot_huber_vs_ridge.py:
```diff
@@ -0,0 +1,65 @@
+decision function approches that of the ridge.
+"""
+
+# Authors: Manoj Kumar mks542@nyu.edu
```
Do we normally have emails here? I can't recall

kastnerkyle commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+    Linear regression model that is robust to outliers.
+
+    The Huber Regressor optimizes the squared loss for the samples where
+    |y - X'w / exp(sigma)| > epsilon and the linear loss for the samples
```
Is linear loss consistent with other linear models notation? Or is absolute loss a better name? We say linear loss in several places but I don't see that terminology in the Owen paper - and in my mind at least linear loss preserves sign

MechCoder replied:
Yes, I think we should call it absolute loss / error. consistent with metrics.mean_absolute_error as well.

kastnerkyle commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,174 @@
+            w = rng.randn(n_features)
+            w[-1] = np.abs(w[-1])
+            grad_same = optimize.check_grad(
+                loss_func, grad_func, w, X, y, 0.01, 0.1)
```
eps here is the default optimize.check_grad argument right? or 0.1? Having trouble reading this

MechCoder replied:
Neither, these are the arguments that are passed to the loss function, i.e epsilon and alpha.

It seems like the step size that it takes to calculate del f / del x is not optional and set to 1e-8.

MechCoder replied:
I'll add the arguments now though :)

kastnerkyle commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+            * exp(sigma))
+    where
+
+        - H(z) = z**2 when |z| < epsilon
```
z**2 -> z ** 2

kastnerkyle commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+
+    More specifically the loss function optimized is.
+        X.shape[0] * exp(sigma) + (
+            \sum_{i=1}^{i=X.shape[0]}H(y_{i} - X_{i}*w - c / exp(sigma))
```
X{i} * w

kastnerkyle commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+    outliers_mask = (abs_linear_loss / sigma) > epsilon
+
+    # Calculate the linear loss due to the outliers.
+    # This is equal to (2 * M * |y - X'w -c / sigma| - M**2)*sigma
```
This is beyond nitpicky... but M\*\*2 -> M \* \* 2) * sigma

kastnerkyle commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+            sigma * n_outliers * epsilon**2)
+
+    # Calculate the quadratic loss due to the non-outliers.-
+    # This is equal to |(y - X'w - c)**2 / sigma**2|*sigma
```
M\*\*2 -> M \* \* 2) * sigma again

kastnerkyle commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+    X_outliers_neg = _axis0_safe_slice(X, outliers_neg, num_outliers_neg)
+
+    if sample_weight is None:
+
```
random space?

kastnerkyle commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,174 @@
+
+    # Check using optimize.check_grad that the gradients are equal.
+    for i in range(5):
+
```
random space, pt 2. ?

kastnerkyle commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+from sklearn.linear_model.base import center_data, LinearModel
+from sklearn.preprocessing import StandardScaler
+from sklearn.utils import (
+    check_X_y, check_array, check_consistent_length, column_or_1d, safe_mask)
```
can we split this into multiple lines

kastnerkyle commented on examples/linear_model/plot_robust_fit.py:
```diff
@@ -24,12 +24,19 @@
 
-from sklearn import linear_model, metrics
+from sklearn.linear_model import (
+  LinearRegression, TheilSenRegressor, RANSACRegressor, HuberRegressor)
```
split into multiple import lines

agramfort commented on examples/linear_model/plot_huber_vs_ridge.py:
```diff
@@ -0,0 +1,65 @@
+plt.plot(x, coef_, 'g-', label="ridge regression")
+
+plt.axhline(y.mean(), color='black')
+plt.axvline(X.mean(), color='black')
```
I'd remove these lines. They don't appear in legend so we don't get it from the plot.

agramfort commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+from math import exp
```
no header. Add authors + license

agramfort replied:
I'll send you a PR. Wait a second.

MechCoder replied:
sure.

agramfort commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+from sklearn.preprocessing import StandardScaler
+from sklearn.utils import (
+    check_X_y, check_array, check_consistent_length, column_or_1d, safe_mask)
+from sklearn.utils.extmath import safe_sparse_dot
```
use relative imports

agramfort commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+
+        # Summing along any axis of a matrix returns a matrix object.
+        outliers_pos_sum = np.squeeze(np.array(X_outliers_pos.sum(axis=0)))
+        outliers_neg_sum = np.squeeze(np.array(X_outliers_neg.sum(axis=0)))
```
the two lines above look fishy

MechCoder replied:
That's what I would have thought as well.. But summing up a sparse matrix returns a sparse matrix, rather than an array. So we convert it into a 2-D array and squeeze it into 1-D.

    X_csr = sparse.rand(5, 5, format='csc')
    X_csr.sum(axis=0)
    matrix([[0., 0., 0., 0., 0.]])

agramfort commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+            weighted_sum = (
+                _dia_matrix(sw_outliers_pos) *
+                X_outliers_pos).sum(axis=0)
+            weighted_sum = np.squeeze(np.array(weighted_sum))
```
same here.

first weighted_sum is already an array and second squeeze is evil

MechCoder replied:
I wish:

https://github.com/numpy/numpy/issues/4780

agramfort commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+            weighted_sum = (
+                _dia_matrix(sw_outliers_neg) *
+                X_outliers_neg).sum(axis=0)
+            weighted_sum = np.squeeze(np.array(weighted_sum))
```
same rmk

agramfort commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+    |y - X'w / exp(sigma)| > epsilon and the linear loss for the samples
+    where |y - X'w / exp(sigma)| < epsilon, where w and sigma are parameters
+    to be optimized. The parameter sigma makes sure that if y is scaled up
+    or down by a certain factor, one does not need to rescale M to acheive
```
what is M?

MechCoder replied:
M -> Manoj coefficient :p

MechCoder commented:
@kastnerkyle @agramfort Thanks for your comments ! I shall address them in a while. I have already addressed a few inline.

agramfort commented:
PR sent https://github.com/MechCoder/scikit-learn/pull/1

have a look. We'll polish this tomorrow.

:beer: time here !

MechCoder commented:
nice ! miss being at the sprint this time ):

MechCoder commented:
addressed your comments and replaced the squeeze with indexing the first element when X is sparse.

MechCoder commented:
Sigh. I'm unable to reproduce Python 2's error with the same SciPy (0.9) which relates the creation of an empty dia_matrix.

agramfort commented on doc/modules/linear_model.rst:
```diff
@@ -1019,6 +1019,62 @@ considering only a random subset of all possible combinations.
+
+.. math::
+
+   \underset{w}{min\,} {n_samples*\sigma + H_m(\frac{X w - y}{\sigma})*\sigma + \alpha {||w||_2}^2}
```
this is not correct as you optimize wrt to \sigma too. Also in the code docstring it's exp(\sigma) in the cost function. What is true?

MechCoder replied:
sorry, its just sigma.

agramfort commented on doc/modules/linear_model.rst:
```diff
@@ -1019,6 +1019,62 @@ considering only a random subset of all possible combinations.
+
+.. topic:: References:
+
+    .. [#f1] Peter J. Huber, Elvezio M. Ronchetti: `Robust Statistics. <http://sanghv.com/download/Soft/Machine%20Learning,%20Artificial%20Intelligence,%20Mathematics%20eBooks/math/statistics/robust%20statistics%20%282nd,%202009%29.pdf>`_
```
this link is not really ok. The book in not public domain... I'd put the doi

agramfort commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,347 @@
+                (self.coef_, [self.intercept_, self.scale_]))
+
+        bounds = np.tile([-np.inf, np.inf], (self.coef_.shape[0], 1))
+        bounds[-1][0] = 1.0
```
maybe add a comment to explains these bounds

agramfort replied:
ping

agramfort commented:
travis is not happy

MechCoder commented:
@agramfort The travis build error is due to some old version of scipy failing to produce an empty matrix. Will fix it up.

Meanwhile it will be useful if you could give another pass, I'll address them this evening :)

agramfort commented:
that's it for me. Make Travis happy and I'll be happy :)

agramfort commented:
MRG+1

MechCoder commented:
Second reviews anyone?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+            w = rng.randn(n_features)
+            w[-1] = np.abs(w[-1])
+            grad_same = optimize.check_grad(
+                loss_func, grad_func, w, X, y,  0.01, 0.1)
```
named arguments maybe? @kastnerkyle and me don't remember the function signature ;)

amueller commented:
Please add huber to this section: http://scikit-learn.org/dev/modules/linear_model.html#robustness-regression-outliers-and-modeling-errors in particular to the trade-off section: Trade-offs: which estimator?

MechCoder commented:
@amueller Addressed your comments.

MechCoder commented:
I think this is ready for another round of reviews @kastnerkyle @amueller

amueller commented:
squash?

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -903,15 +903,21 @@ in these settings.
 
-   * :ref:`RANSAC <ransac_regression>` is faster, and scales much better
-     with the number of samples
+  * :ref:`HuberRegressor` <huber_regression> should be faster than
```
I would reformulate these two sentences. First you say something, then you say it is only sometimes true.
What are the trade-offs? HuberRegressor should be fast with < X samples? When n_samples << n_features?

MechCoder replied:
I wish it would be so easy to predict heuristics for what `n_samples` I should use what estimator. This depends on HUGE number of factors. 

RANSAC would definitely be faster for `n_samples` >> `n_features` since it produces 100 fits but it is likely to be less robust because of the smaller number of samples doing each fit. The same argument goes to TheilSenRegressor because the smaller your `n_subsamples` is, it is going to be less robust.

We could just say that TheilSenRegressor and RANSAC are to be faster but they are unlikely to be as robust as HuberRegressor for the default parameters when n_samples >> n_features. How does that sound?

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -1051,6 +1057,62 @@ considering only a random subset of all possible combinations.
+Huber Regression
+----------------
+
+The :class:`HuberRegressor` is an extension of :class:`Ridge` except that
```
I wouldn't say it is an extension. Maybe say it is similar in that it also applies L2 regularization and a squared loss for "inliers"?

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -1051,6 +1057,62 @@ considering only a random subset of all possible combinations.
+
+.. math::
+
+   \underset{w, \sigma}{min\,} {n_samples*\sigma + H_m(\frac{X w - y}{\sigma})*\sigma + \alpha {||w||_2}^2}
```
have you checked that this renders nicely?

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -1051,6 +1057,62 @@ considering only a random subset of all possible combinations.
+
+.. math:: 2*M*|z| - M^2 if |z| >= \epsilon
+
+It is advised to set the parameter `epsilon` to 1.35 to acheive 95% statistical efficiency.
```
typo in achieve. Maybe say [whatever the reference is] advises to set the parameter epsilon to....

Is epsilon a good name?

MechCoder replied:
I copied the name from the `SGDRegressor` where `epsilon` does the same thing.

amueller commented:
I'm not an expert in robust regression. I think it would be good if we could check our parametrization and implementation against R, so that R users can more or less easily transition to our estimator, and we have an additional confirmation of the implementation [adding a unit test against a reference implementation would be nice, I think].

amueller commented:
There is a nice intro here: http://www.ats.ucla.edu/stat/r/dae/rreg.htm

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -1051,6 +1057,62 @@ considering only a random subset of all possible combinations.
+The :class:`HuberRegressor` differs from using :class:`SGDRegressor` with loss set to `huber`
+in the following ways.
+
+- :class:`HuberRegressor` is scaling invariant. Once `epsilon` is set, scaling X and y
```
double ticks around epsilon. (maybe around X and y , too)

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -1051,6 +1057,62 @@ considering only a random subset of all possible combinations.
+  scaled.
+
+- :class:`HuberRegressor` should be more efficient to use on data with small number of
+  samples while :class:`SGDRegressor` needs a number of passes on the training data to
```
This seems a slightly odd distinction. HuberRegressor makes more passes over the data, right? SGDRegressor scales better to larger data but takes longer on small data?

MechCoder replied:
Yes ! When we have small number of samples, we need a large number of passes to acheive the same robustness as the HuberRegressor.

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -1051,6 +1057,62 @@ considering only a random subset of all possible combinations.
+
+    .. [#f1] Peter J. Huber, Elvezio M. Ronchetti: Robust Statistics, Concomitant scale estimates, pg 172
+
+    .. [#f2] Art B. Owen: `A robust hybrid of lasso and ridge regression. <http://statweb.stanford.edu/~owen/reports/hhu.pdf>_`
```
How is this relevant again?

amueller commented on examples/linear_model/plot_huber_vs_ridge.py:
```diff
@@ -0,0 +1,66 @@
+Fit Ridge and HuberRegressor on a dataset with outliers.
+
+The example shows that the coefficients in ridge are strongly influenced
+by the outliers are added to the dataset. The Huber regressor is less
```
I would use "present" rather than added, as no-one adds outliers intentionally ;).
The sentence currently doesn't parse. Either "when outliers are present" or "by the outliers that are present".

amueller commented on examples/linear_model/plot_huber_vs_ridge.py:
```diff
@@ -0,0 +1,66 @@
+
+The example shows that the coefficients in ridge are strongly influenced
+by the outliers are added to the dataset. The Huber regressor is less
+influenced by the outliers since the model fits the linear loss for these.
```
I don't think you fit a loss. Uses a linear loss?

amueller commented on examples/linear_model/plot_huber_vs_ridge.py:
```diff
@@ -0,0 +1,66 @@
+by the outliers are added to the dataset. The Huber regressor is less
+influenced by the outliers since the model fits the linear loss for these.
+As the parameter epsilon is increased for the Huber regressor, the decision
+function approches that of the ridge.
```
approaches

amueller commented on examples/linear_model/plot_robust_fit.py:
```diff
@@ -22,14 +22,20 @@
-  a break point above which it performs worst than OLS.
+  a break point above which it performs worse than OLS.
+
+- HuberRegressor should not differ much in performance due to outliers in
```
differ much from what? TheilSen? OLS?

amueller commented on sklearn/linear_model/__init__.py:
```diff
@@ -18,6 +18,7 @@
                                  lasso_path, enet_path, MultiTaskLasso,
                                  MultiTaskElasticNet, MultiTaskElasticNetCV,
                                  MultiTaskLassoCV)
+from .huber import HuberRegressor
```
add to ``__all__``?

amueller commented on examples/linear_model/plot_robust_fit.py:
```diff
@@ -55,13 +61,15 @@
-              ('RANSAC', linear_model.RANSACRegressor(random_state=42)), ]
-colors = {'OLS': 'turquoise', 'Theil-Sen': 'gold', 'RANSAC': 'lightgreen'}
-linestyle = {'OLS': '-', 'Theil-Sen': '-.', 'RANSAC': '--'}
+plt.clf()
```
why?

amueller commented on examples/linear_model/plot_huber_vs_ridge.py:
```diff
@@ -0,0 +1,66 @@
+
+# Fit the huber regressor over a series of epsilon values.
+colors = ['r-', 'b-', 'y-', 'm-']
+markers = ['v', 'o', 's', 'D']
```
what are the markers for? color blind people / printing? It kinda looks like they mark some points on the lines, which I don't think they do.

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+from .base import LinearModel
+from ..utils import check_X_y
+from ..utils import check_consistent_length
+from ..utils import column_or_1d
```
unused import

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+
+def _dia_matrix(X):
+    """
+    Old versions of scipy cannot create an empty dia_matrix
```
Can you give the exact version?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+    and non_outliers are typically non-zero and it makes the code
+    tougher to follow.
+    """
+    if len_mask != 0:
```
do we maybe want to change that in ``safe_mask``?

MechCoder replied:
maybe but preferably not in this PR

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+        Target vector.
+
+    epsilon: float
+        Measures the robustness of the Huber estimator.
```
It doesn't measure, this is a parameter, right?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+    grad[-1] -= n_outliers * epsilon ** 2
+    grad[-1] -= squared_loss / sigma
+
+    # Gradient due to the intercept.
```
this is gradient wrt the intercept, right? And same for sigma above?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+rng = np.random.RandomState(0)
+
+
+def make_regression_with_outliers(n_samples=50, n_features=20):
```
Maybe add a one-line comment explaining the function. the data has outliers in X, not in y, right?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+    ridge.fit(X, y)
+    huber = HuberRegressor(fit_intercept=True, epsilon=1e2, alpha=0.0)
+    huber.fit(X, y)
+    assert_almost_equal(huber.coef_, ridge.coef_, 2)
```
why only 2? Are the solvers struggling because of alpha=0? Shouldn't we use ``LinearRegression`` instead of ``Ridge`` then?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+
+
+def test_huber_equals_ridge_for_high_epsilon():
+    """Test that Ridge matches HuberRegressor for large epsilon"""
```
please use comments, not docstrings.

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+
+
+def test_huber_gradient():
+    """Test that the gradient calculated by _huber_loss_and_gradient is ok"""
```
please use comments, not docstrings.

amueller replied:
maybe correct instead of ok?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+
+
+def test_huber_sample_weights():
+    """Test sample_weights implementation in HuberRegressor"""
```
please use comments, not docstrings (also below).

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+    assert_array_almost_equal(huber.intercept_, huber_intercept, 3)
+
+    # Test sparse implementation with sparse weights.
+    # Checking sparse=non_sparse should be covered in the common tests.
```
Is it? I don't think it is.

MechCoder replied:
See https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/estimator_checks.py#L327 . Git blame tells me you wrote it, but I don't blame you for forgetting though ;)

amueller replied:
It doesn't check that the result is the same, does it?

MechCoder replied:
Oh but we should, shouldn't we?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+    ----------
+    epsilon: float, greater than 1.0, default 1.35
+        The parameter epsilon controls the number of samples that should be
+        classified as outliers. The lesser the epsilon, the more robust it is
```
The smaller epsilon, the more robust the fit is to outliers.

amueller replied:
Or maybe "The smaller epsilon, the more points will be declared outliers"? Or "Smaller epsilon leads to more points being declared outliers". What does ``epsilon=1`` mean? And why does it need to be larger than 1?
Thats not really clear from the formula. Though maybe an explanation should go into the user guide?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+
+    Parameters
+    ----------
+    epsilon: float, greater than 1.0, default 1.35
```
always space before ``:``

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+    outliers while not completely ignoring their effect.
+
+    More specifically the loss function optimized is.
+        n_samples * exp(sigma) + (
```
I think you need a ``math`` block here. Please check if sphinx complains and look at the rendering in html (and preferrably in pdf, too).

MechCoder replied:
I think I'll remove this because it is already present in the User Guide.

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+    """Linear regression model that is robust to outliers.
+
+    The Huber Regressor optimizes the squared loss for the samples where
+    |(y - X'w) / exp(sigma)| < epsilon and the absolute loss for the samples
```
pretty sure the math needs to be in double backticks to not break sphinx.

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+            * sigma)
+    where
+
+        - H_m(z) = z**2 when |z| < epsilon
```
math:: ? or double backticks?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+    assert_array_almost_equal(huber_sparse.coef_, huber_coef, 3)
+
+
+def return_number_outliers(X, y, coef, intercept, scale, epsilon):
```
why don't you pass ``huber`` and use ``predict(X) - y > epsilon * scale``?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+        X_new, y, huber.coef_, huber.intercept_, huber.scale_, huber.epsilon)
+
+    assert_equal(n_outliers1, n_outliers2)
+    assert_equal(n_outliers3, n_outliers4)
```
I feel the logic would be simpler if you compare 3 to 1.

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+def return_number_outliers(X, y, coef, intercept, scale, epsilon):
+    """Return the number of outliers."""
+    outliers = np.abs(np.dot(X, coef) + intercept - y) > epsilon * scale
+    return np.sum(outliers)
```
is there a reason to compare the number instead of which points are outliers? That would be a stronger test, right?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+        alpha=0.0, loss="huber", shuffle=True, random_state=0, n_iter=1000000,
+        fit_intercept=False, epsilon=1.35)
+    sgdreg.fit(X_scale, y_scale)
+    assert_array_almost_equal(huber.coef_, sgdreg.coef_, 2)
```
why only 2? you seem to have a lot of iterations.

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+    assert_array_almost_equal(huber.coef_, sgdreg.coef_, 2)
+
+
+def test_huber_warm_start():
```
I don't undestand this test. The first fit does 10 iterations, the second 5, right?
Can you add a comment explaining the test?

MechCoder replied:
changed.

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+    X, y = make_regression_with_outliers()
+    huber = HuberRegressor(fit_intercept=True, alpha=0.01, n_iter=100)
+    huber.fit(X, y)
+    linear_loss = np.dot(X, huber.coef_) + huber.intercept_ - y
```
why not ``predict``?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+    huber = HuberRegressor(fit_intercept=True, alpha=0.01, n_iter=100)
+    huber.fit(X, y)
+    linear_loss = np.dot(X, huber.coef_) + huber.intercept_ - y
+    mask = np.abs(linear_loss) < huber.epsilon * huber.scale_
```
``inlier_mask`` ?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+    ridge.fit(X, y)
+    ridge_score = ridge.score(X[mask], y[mask])
+    ridge_outlier_score = ridge.score(X[~mask], y[~mask])
+    assert_greater(huber_score, ridge_score)
```
why? I don't understand this.

MechCoder replied:
clarified IRL

amueller replied:
Maybe add a comment?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,180 @@
+
+
+def test_huber_better_r2_score():
+    """Test that huber returns a better r2 score than non-outliers"""
```
This explanation is unclear to me.

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+        Regularization parameter.
+
+    warm_start: bool, default False
+        If warm_start is set to False, the coefficients will be overwritten
```
Is this the standard docstring? Did you copy & paste from logistic/ridge?

MechCoder replied:
Yes. But I'll change it.

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+        if the data is already centered around the origin.
+
+    pgtol: float, default 1e-5
+        The iteration will stop when max{|proj g_i | i = 1, ..., n} <= pgtol
```
double ticks around math or use a math expression.

amueller replied:
why not just ``tol``? that is the standard name, isn't it?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+
+    Attributes
+    ----------
+    coef_: array, shape (n_features,)
```
space before column

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+        Features got by optimizing the Huber loss.
+
+    intercept_: float
+        Bias factor.
```
It's not a factor, is it?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+        Bias factor.
+
+    scale_ : float
+        The value by which |y - X'w - c| is scaled down by.
```
remove last "by", double ticks around math.

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+        The value by which |y - X'w - c| is scaled down by.
+
+    n_iter_: int
+        Number of iterations that fmin_l_bfgs_b has run for, if available.
```
when would it not be available?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+           http://statweb.stanford.edu/~owen/reports/hhu.pdf
+    """
+
+    def __init__(self, epsilon=1.35, n_iter=100, alpha=0.0001,
```
where do ``n_iter`` and ``pgtol`` defaults come from?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+            sample_weight = np.array(sample_weight)
+            check_consistent_length(X, y, sample_weight)
+        else:
+            check_consistent_length(X, y)
```
this is already done in ``check_X_y``

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+
+    This loss function is jointly convex in sigma and w and hence can be
+    optimized together.
+
```
Link to the user guide here.

amueller commented on examples/linear_model/plot_huber_vs_ridge.py:
```diff
@@ -0,0 +1,66 @@
+
+Fit Ridge and HuberRegressor on a dataset with outliers.
+
+The example shows that the coefficients in ridge are strongly influenced
```
it shows the predictions / fit, not the coefficients, right?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+            else:
+                self.coef_ = np.zeros(X.shape[1] + 1)
+        elif self.warm_start:
+            self.coef_ = np.concatenate(
```
this seems very odd. the meaning of ``coef_`` changes within this function. why do you need to store the full coefficient vector in ``self``? couldn't you just call this "parameters" and have it be a local variable and pass it to lbfgs?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+            raise ValueError(
+                "epsilon should be greater than 1.0, got %f" % self.epsilon)
+        coef = getattr(self, 'coef_', None)
+        if not self.warm_start or (self.warm_start and coef is None):
```
the ``self.warm_start and`` is vacuous, right?
Are you using coef anywhere else?
Why not
```
if self.warm_start and hasattr(self, coef_):
    parameters = np.concatenate((self.coef_, [self.intercept_, self.scale_]))
else:
    if self.fit_intercept:
        parameters = ...
    else:
        parameters = ...
```

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+
+        try:
+            self.coef_, f, self.dict_ = optimize.fmin_l_bfgs_b(
+                _huber_loss_and_gradient, self.coef_, approx_grad=True,
```
why is ``approx_grad=True``?
And when is a ``TypeError`` raised?

MechCoder replied:
TypeError is caused in old versions of scipy because of no maxiter argument. Will add a comment

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,349 @@
+    # Calculate the linear loss due to the outliers.
+    # This is equal to (2 * M * |y - X'w -c / sigma| - M**2) * sigma
+    outliers = abs_linear_loss[outliers_mask]
+    num_outliers = np.count_nonzero(outliers)
```
I don't think it's great if there is ``num_outliers`` and ``n_outliers`` and they have quite different semantics (which are not explained). Actually I'm not very sure what happens here. ``num_outliers`` is the number of outliers with non-zero loss, right? But the loss is float. How can it be zero?

MechCoder replied:
`num_outliers` and `n_outliers` are the same except that `n_outliers` are the weighted number including sample_weights.

Yikes! I meant to write `np.count_nonzero(outliers_mask)` . Let me fix it and pray that it does not break anything.

amueller commented:
I didn't go through the computation of the objective and the gradient yet.
I feel maybe there is a chance for that to be slightly refactored, but I'm not sure. Clearing up ``num_outliers`` vs ``n_outliers`` would help. Also, why is it necessary to separate positive and negative parts of the loss? Wouldn't it be easier to just store a sign?
Finally: are there significant gains in performance by special-casing the ``sample_weights = np.ones(n_samples)`` case?

MechCoder commented:
@amueller Thanks for your comments. Will address them on Monday. Hopefully we can get this merged before the sem break !

MechCoder commented:
> Wouldn't it be easier to just store a sign?

Yes, pity I overlooked that. Working on the refactor.

MechCoder commented:
@amueller I have addressed your comments. It might not be correct it to compare it with the R implementation because it does not do the same thing. It uses the closed form solution gives it a weight of one if the residual is less than a K , and K / residual if the residual is greater than a certain value.

MechCoder commented:
@amueller Squashed and rebased. Would be great if you could have another look.

MechCoder commented:
You were right about not gaining much about the sample_weights case. I shall remove those.

MechCoder commented:
Done. This is ready for another round of reviews. (Hopefully final.)

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -1050,6 +1059,67 @@ considering only a random subset of all possible combinations.
+The :class:`HuberRegressor` is similar to :class:`Ridge` that it also applies
+L2 regularization and a squared loss for samples which are classified as inliers.
+A sample is classified as an inlier if the absolute error of that sample is
+lesser than a certain threshold. It differs from :class:`TheilSenRegressor`
```
This reads weirdly. "It's similar to Ridge because X" "It's different from TheilSen because Y". I would first expect to hear what the difference with Ridge is!

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -1050,6 +1059,67 @@ considering only a random subset of all possible combinations.
+
+- :class:`HuberRegressor` is scaling invariant. Once ``epsilon`` is set, scaling ``X`` and ``y``
+  down or up by different values would produce the same robustness to outliers as before.
+  as compared to :class:`SGDRegessor` where `epsilon` has to be set again when ``X`` and ``y`` are
```
double backticks around epsilon

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -902,15 +902,24 @@ in these settings.
-     with the number of samples
+  * :ref:`HuberRegressor <huber_regression>` should be faster than
+    :ref:`RANSAC <ransac_regression>` and :ref:`Theil Sen <theil_sen_regression>`
+    unless the number of samples are very large, i.e n_samples >> n_features.
```
nitpick: double backticks around n_samples and n_features

amueller commented on examples/linear_model/plot_robust_fit.py:
```diff
@@ -22,14 +22,20 @@
-  a break point above which it performs worst than OLS.
+  a break point above which it performs worse than OLS.
+
+- HuberRegressor should not differ much in performance to both RANSAC
```
this seems to be slightly in conflict in what the user guide says. also: is performance here accuracy or training time?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,267 @@
+
+    Parameters
+    ----------
+    w: ndarray, shape (n_features + 1,) or (n_features + 2,)
```
nitpick: space in front of  ``:`` for consistency

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,267 @@
+    ----------
+    w: ndarray, shape (n_features + 1,) or (n_features + 2,)
+        Feature vector.
+        w[:n_features] gives the feature vector
```
I would say "is the" not "gives the". Also: feature vector? you mean coefficient, right?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,267 @@
+    loss: float
+        Huber loss.
+
+    gradient: ndarray, shape (n_features + 1,) or (n_features + 2,)
```
maybe ``shape len(w)`` ?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,267 @@
+    ``|(y - X'w) / sigma| < epsilon`` and the absolute loss for the samples
+    where ``|(y - X'w) / sigma| > epsilon``, where w and sigma are parameters
+    to be optimized. The parameter sigma makes sure that if y is scaled up
+    or down by a certain factor, one does not need to rescale epsilon to acheive
```
achieve

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,267 @@
+    ----------
+    epsilon : float, greater than 1.0, default 1.35
+        The parameter epsilon controls the number of samples that should be
+        classified as outliers. The lesser the epsilon, the more robust it is
```
The smaller

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,267 @@
+        to outliers.
+
+    max_iter : int, default 100
+        Number of iterations that scipy.optimize.fmin_l_bfgs_b should run for.
```
Maximum number of iterations that...

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,267 @@
+        if the data is already centered around the origin.
+
+    tol : float, default 1e-5
+        The iteration will stop when max{|proj g_i | i = 1, ..., n} <= tol
```
How does that look in sphinx? maybe double backticks? Please check the rendering, ok? Maybe you want math?

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,267 @@
+    scale_ : float
+        The value by which ``|y - X'w - c|`` is scaled down.
+
+    n_iter_: int
```
space before :

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,267 @@
+        y : array-like, shape (n_samples,)
+            Target vector relative to X.
+
+        sample_weight: array-like, shape (n_samples,)
```
space before :

amueller commented on sklearn/linear_model/huber.py:
```diff
@@ -0,0 +1,267 @@
+
+        if self.epsilon < 1.0:
+            raise ValueError(
+                "epsilon should be greater than 1.0, got %f" % self.epsilon)
```
greater or equal?

amueller commented on sklearn/linear_model/tests/test_huber.py:
```diff
@@ -0,0 +1,176 @@
+    assert_array_almost_equal(huber_sparse.coef_, huber_coef, 3)
+
+
+def return_outliers(X, y, huber):
```
do we want to add an attribute to the estimator for that?

amueller commented on sklearn/utils/__init__.py:
```diff
@@ -60,6 +60,26 @@ def safe_mask(X, mask):
     return mask
 
 
+def axis0_safe_slice(X, mask, len_mask):
```
have we reimplemented this in RANSAC?

MechCoder replied:
RANSAC raises an error if all the data points are classified as outliers. But in this case we don't have a fixed threshold, the term `M*sigma` varies for each iteration

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1486,7 +1486,11 @@ def check_non_transformer_estimators_n_iter(name, estimator,
-    assert_greater(estimator.n_iter_, 0)
+
+    # HuberRegressor depends on scipy.optimize.fmin_l_bfgs_b
+    # which does return a n_iter for old versions of SciPy.
```
does or doesn't?

amueller commented:
lgtm apart from nitpicks. Maybe adding an attribute that stores which points are outliers on the training set would be interesting.

amueller commented:
feel free to merge once tests pass. we can always add a an attribute for the outliers.
we do need an entry to whatsnew and a "versionadded" tag.

MechCoder commented:
I have 3 more minor todos

1. Add a property for the outliers
2. Check the gradient doc rendering
3. Modify the documentation of the example.

Will address in a while

agramfort commented:
ping us when we should look again

MechCoder commented:
@agramfort done!!

MechCoder commented:
Tests pass!! Merging with master :D

agramfort commented:
great work @MechCoder !

amueller commented:
thanks @MechCoder ! :beers:

GaelVaroquaux commented:
> thanks @MechCoder ! :beers:

Yes. Awesome!!

AndreasTheodoulou commented:
Hi all. I am currently using HuberRegressor and I find that the results tend to change a lot (much worse), when I increase the regularization more than the default value (e.g alpha = 0.0001221677 from alpha = 0.001), and when I change the fit_intercept from True to False.

Any ideas why that might be the case and if there is any way to tackle this?
 
Note: my main concern for my use case is using fit_intercept = False, but being able to test regularization would also be nice
